runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcDirectGetWorkStream.java

```diff
@@ -0,0 +1,346 @@
+      StreamObserverFactory streamObserverFactory,
+      Set<AbstractWindmillStream<?, ?>> streamRegistry,
+      int logEveryNStreamFailures,
+      ThrottleTimer getWorkThrottleTimer,
+      Supplier<GetDataStream> getDataStream,
+      Supplier<CommitWorkStream> commitWorkStream,
+      ProcessWorkItem processWorkItemFn) {
+    super(
+        startGetWorkRpcFn, backoff, streamObserverFactory, streamRegistry, logEveryNStreamFailures);
+    this.request = request;
+    this.getWorkThrottleTimer = getWorkThrottleTimer;
+    this.processWorkItemFn = processWorkItemFn;
+    this.workItemBuffers = new ConcurrentHashMap<>();
+    // Use the same GetDataStream and CommitWorkStream instances to process all the work in this
+    // stream.
+    this.getDataStream = Suppliers.memoize(getDataStream::get);
+    this.commitWorkStream = Suppliers.memoize(commitWorkStream::get);
+
+    this.inflightMessages = new AtomicLong();
+    this.inflightBytes = new AtomicLong();
+    this.nextBudgetAdjustment = new AtomicReference<>(GetWorkBudget.noBudget());
+    this.pendingResponseBudget = new AtomicReference<>(GetWorkBudget.noBudget());
+  }
+
+  public static GrpcDirectGetWorkStream create(
+      Function<
+              StreamObserver<StreamingGetWorkResponseChunk>,
+              StreamObserver<StreamingGetWorkRequest>>
+          startGetWorkRpcFn,
+      GetWorkRequest request,
+      BackOff backoff,
+      StreamObserverFactory streamObserverFactory,
+      Set<AbstractWindmillStream<?, ?>> streamRegistry,
+      int logEveryNStreamFailures,
+      ThrottleTimer getWorkThrottleTimer,
+      Supplier<GetDataStream> getDataStream,
+      Supplier<CommitWorkStream> commitWorkStream,
+      ProcessWorkItem processWorkItemFn) {
+    GrpcDirectGetWorkStream getWorkStream =
+        new GrpcDirectGetWorkStream(
+            startGetWorkRpcFn,
+            request,
+            backoff,
+            streamObserverFactory,
+            streamRegistry,
+            logEveryNStreamFailures,
+            getWorkThrottleTimer,
+            getDataStream,
+            commitWorkStream,
+            processWorkItemFn);
+    getWorkStream.startStream();
+    return getWorkStream;
+  }
+
+  private void sendRequestExtension() {
+    GetWorkBudget adjustment = nextBudgetAdjustment.get();
+    StreamingGetWorkRequest extension =
+        StreamingGetWorkRequest.newBuilder()
+            .setRequestExtension(
+                Windmill.StreamingGetWorkRequestExtension.newBuilder()
+                    .setMaxItems(adjustment.items())
+                    .setMaxBytes(adjustment.bytes()))
+            .build();
+
+    executor()
+        .execute(
+            () -> {
+              try {
+                send(extension);
+              } catch (IllegalStateException e) {
+                // Stream was closed.
+              }
+            });
+
+    synchronized (this) {
+      // Just sent the request extension, reset the nextBudgetAdjustment. This will be set when
```

Comment:
I think it would be better to in the same synchronized block take the budget and clear it. Otherwise it seems possibly racy and this could lose budget.

Reply:
done